### PR TITLE
get num tissues

### PR DESCRIPTION
**Pull request by nanli-emory**
feature added: # of pieces of tissue [#276](https://github.com/choosehappy/HistoQC/issues/276)

1.  add helper function `BaseImage.getMaskReginsStats` to get the based statistics for tissue areas
2. Use `BasicModule.countTissuePieces` to get # tissues in the current mask
3. change the `./histoqc/config/config_v2.1.ini` files to save `#pieces_of_tissue`

### CODE REVIEW

File: histoqc/BasicModule.py [hunk 1]
```diff
@@ -69,3 +69,10 @@ def finalProcessingArea(s, params):
             f"{s['filename']} - After BasicModule.finalProcessingArea NO tissue remains detectable! Downstream modules likely to be incorrect/fail")
         s["warnings"].append(
             f"After BasicModule.finalProcessingArea NO tissue remains detectable! Downstream modules likely to be incorrect/fail")
+
+
+def countTissuePieces(s, params):
+    area_thresh = int(params.get("area_threshold", "1000"))
```
Comment (jacksonjacobs1) [inline histoqc/BasicModule.py]: I think applying an area threshold here is redundant. Ideally small objects etc. will already be filtered out and the `countTissuePieces` method will just report what's there.

File: histoqc/BlurDetectionModule.py [hunk 2]
```diff
@@ -37,19 +37,20 @@ def identifyBlurryRegions(s, params):
...
-        nobj = area_max = area_mean = 0
+    # rps = measure.regionprops(morphology.label(mask))
+    # if rps:
+    #     areas = np.asarray([rp.area for rp in rps])
+    #     nobj = len(rps)
+    #     area_max = areas.max()
+    #     area_mean = areas.mean()
+    # else:
+    #     nobj = area_max = area_mean = 0
+    stat_rs = getMaskReginsStats(mask)
```
Comment (jacksonjacobs1) [inline histoqc/BlurDetectionModule.py]: Looks like none of the stats are being added to the printList?

File: histoqc/BaseImage.py [hunk 3]
```diff
@@ -341,3 +342,22 @@ def getDimensionsByOneDim(s: BaseImage, dim: int) -> Tuple[int, int]:
     else:
         w = int(dim * width / height)
         return w, dim
+
+
+def getMaskReginsStats(img, area_threshold: int = 0)-> dict:
```
Comment (jacksonjacobs1) [inline histoqc/BaseImage.py]: regions misspelled

File: histoqc/BaseImage.py [hunk 4]
```diff
@@ -206,6 +206,9 @@ def getImgThumb(self, size: str):
                 self[key] = getBestThumb(self, bx, by, target_dims, target_sampling_factor)
         return self[key]
 
+'''
+the followings are helper functions
```
Comment (jacksonjacobs1) [inline histoqc/BaseImage.py]: "following" not "followings"

**Comment by jacksonjacobs1** [comment]
Thanks Nan, 

Just a few small comments above.

**Comment by nanli-emory** [comment]
> Thanks Nan,
> 
> Just a few small comments above.

Hi @jacksonjacobs1, please review and confirm the changes.

**Review by jacksonjacobs1 — APPROVED** [review]
Looks good! Ready to merge in.

**Comment by jacksonjacobs1** [comment]
@choosehappy one quick question before I merge this in....

This PR adds a line to config v2.1 to count # of tissue pieces.

I do not think we should modify config files while keeping the same filename. Conceptually, histoqc should always produce the same output if the user uses config v2.1. Do you agree?

**Comment by choosehappy** [comment]
that is a very good, even philosophical, question

at minimum I would say the same config file should produce the same numerical results, and in this case, simply counting the pieces of tissue isn't actually impacting the *metrics*

on the other hand, the output file itself now changes - which is less than ideal ,but may be tolerable

how do you want to address this? i think we need config files to be "fully featureD" so that folks can see how to use the different things. on the other hand, "2.1" was for the 2.1 release --- are we still on the 2.1 release if we keep adding things? 

at some point we should officially release (tag + code) e,g. 2.2, 2.3 etc - maybe having a config name based on the version is a bit chaotic and unsustainable? maybe it should all go into just config.ini and have the versioning "do its thing"? 

what do you think?

**Comment by jacksonjacobs1** [comment]
I agree that the config file v2.1 naming convention is a bit chaotic.

However, pushing changes into the config.ini kind of defeats the purpose of nan's unit tests:
https://github.com/choosehappy/HistoQC/blob/master/histoqc/tests/test_images_tsv_results.py

The unit tests were written specifically for a configuration file (in this case v2.1) which we assume does not change between commits.

If we change config files in commits, we will also need to update the unit tests before each commit.

**Comment by choosehappy** [comment]
> If we change config files in commits, we will also need to update the unit tests before each commit.

I don't immediately know that this is a bad thing? if we're adding new features, we should be adding new unit tests to cover those features? 

so it somehow seems like this is the intended outcome?

**Pull request merged**